include/cuco/static_multimap.cuh

```diff
@@ -0,0 +1,1716 @@
+ * Therefore, the objects must have unique, bitwise object representations (e.g., no padding bits).
+ * - Does not support erasing keys
+ * - Capacity is fixed and will not grow automatically
+ * - Requires the user to specify sentinel values for both key and mapped value
+ * to indicate empty slots
+ * - Does not support concurrent insert and find operations
+ *
+ * The `static_multimap` supports two types of operations:
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `insert`, `contains`, `count`, `retrieve` and their
+ * variants. These APIs should be used when there are a large number of keys to insert or lookup in
+ * the map. For example, given a range of keys specified by device-accessible iterators, the bulk
+ * `insert` function will insert all keys into the map.
+ *
+ * The singular device-side operations allow individual threads to perform
+ * independent operations (e.g. `insert`, etc.) from device code. These
+ * operations are accessed through non-owning, trivially copyable "view" types:
+ * `device_view` and `device_mutable_view`. The `device_view` class is an
+ * immutable view that allows only non-modifying operations such as `count` or
+ * `contains`. The `device_mutable_view` class only allows `insert` operations.
+ * The two types are separate to prevent erroneous concurrent insert/find
+ * operations.
+ *
+ * Loading two consecutive slots instead of one for small pairs can improve cache utilization since
+ * 16B memory loads are natively supported at the SASS/hardware level. This `vector load` method is
+ * implicitly applied to all involved operations (e.g. `insert`, `count`, and `retrieve`, etc.) if
+ * the pair is packable (see `multimap::uses_vector_load` logic).
+ *
+ * Example:
+ * \code{.cpp}
+ * int empty_key_sentinel = -1;
+ * int empty_value_sentinel = -1;
+ *
+ * // Constructs a multimap with 100,000 slots using -1 and -1 as the empty key/value
+ * // sentinels. Note the capacity is chosen knowing we will insert 50,000 keys,
+ * // for an load factor of 50%.
+ * static_multimap<int, int> m{100'000, empty_key_sentinel, empty_value_sentinel};
+ *
+ * // Create a sequence of pairs {{0,0}, {1,1}, ... {i,i}}
+ * thrust::device_vector<thrust::pair<int,int>> pairs(50,000);
+ * thrust::transform(thrust::make_counting_iterator(0),
+ *                   thrust::make_counting_iterator(pairs.size()),
+ *                   pairs.begin(),
+ *                   []__device__(auto i){ return thrust::make_pair(i,i); };
+ *
+ *
+ * // Inserts all pairs into the map
+ * m.insert(pairs.begin(), pairs.end());
+ *
+ * // Get a `device_view` and passes it to a kernel where threads may perform
+ * // `contains/count/retrieve` lookups
+ * kernel<<<...>>>(m.get_device_view());
+ * \endcode
+ *
+ *
+ * @tparam Key Arithmetic type used for key
+ * @tparam Value Type of the mapped values
+ * @tparam ProbeSequence Probe sequence used in multimap
```

Comment:
There needs to be more documentation on what the requirements of the ProbeSequence type are, e.g., required member functions/types.